gobblin-runtime/src/main/java/org/apache/gobblin/runtime/GobblinMultiTaskAttempt.java

```diff
@@ -197,12 +199,20 @@ public Void call()
   public void shutdownTasks()
       throws InterruptedException {
     log.info("Shutting down tasks");
-    for (Task task: this.tasks) {
-      task.shutdown();
+    for (Map.Entry<Task, Future<?>> task : this.tasks.entrySet()) {
+      task.getKey().shutdown();
     }
 
-    for (Task task: this.tasks) {
-      task.awaitShutdown(1000);
+    for (Map.Entry<Task, Future<?>> task : this.tasks.entrySet()) {
+      task.getKey().awaitShutdown(1000);
+    }
+    for (Map.Entry<Task, Future<?>> task : this.tasks.entrySet()) {
+      if (task.getValue().cancel(true)) {
+        log.info("Task cancelled.");
```

Comment:
I think you can add the awaitShutdown() here and remove the onTaskRunCompletion(). The task.run() has a finally block that calls onTaskRunCompletion().